File: 345. Reverse Vowels of a String.cpp
class Solution {
public:
    // Two pointers: O(N)time, O(1)space, 82.60%, ok
    string reverseVowels(string s) {
        if(s.empty()) return s;
        // set<char> vowels = {'a', 'A', 'e', 'E', 'i', 'I', 'o', 'O', 'u', 'U'};
        string vowels = "AEIOUaeiou";  // more elegant
        int head = 0, tail = s.size() - 1;
        while(head < tail){
            // while(head < tail && !vowels.count(s[head])) head++; // for set
            // while(head < tail && !vowels.count(s[tail])) tail--;
            while(head < tail && vowels.find(s[head]) == string::npos) head++; // for string
            while(head < tail && vowels.find(s[tail]) == string::npos) tail--;
            swap(s[head], s[tail]);
            head++; tail--;  // bug: should update or stuck in loop
        }
        return s;
    }
};

File: 345. Reverse Vowels of a String.py
class Solution(object):
    
    # Two pointer: O(N) time & space
    def reverseVowels(self, s):
        """
        :type s: str
        :rtype: str
        """
        p1, p2 = 0, len(s)-1
        vowels = ['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U']
        s = [c for c in s]
        while p1 < p2:
            while p1 < len(s) and s[p1] not in vowels: p1 += 1
            while p2 >= 0 and s[p2] not in vowels: p2 -= 1
            if 0 <= p1 < p2 < len(s): 
                s[p1], s[p2] = s[p2], s[p1]
                p1, p2 = p1+1, p2-1
        return ''.join(s)
        
        # regular expression: 